### PR TITLE
Add dbus methods: deleteContact and deleteRecipient

### PR DESCRIPTION
**Pull request by JtheSaw**
- implement deleteContact
- implement deleteRecipient
- Fix null pointer exceptions
- Update documentation

### CODE REVIEW

File: lib/src/main/java/org/asamk/signal/manager/helper/ProfileHelper.java [hunk 1]
```diff
@@ -194,6 +194,9 @@ private Profile getRecipientProfile(RecipientId recipientId, boolean force) {
         }
 
         profile = decryptProfileIfKeyKnown(recipientId, encryptedProfile);
+        if (profile.getGivenName() == null && profile.getFamilyName() == null) {
```
Comment (AsamK) [inline lib/src/main/java/org/asamk/signal/manager/helper/ProfileHelper.java:197]: why was this added? Even if the names are null the other parts of the profile should still be stored.

File: lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java [hunk 2]
```diff
@@ -261,7 +261,12 @@ public ProfileKeyCredential getProfileKeyCredential(final RecipientId recipientI
     public void storeProfile(final RecipientId recipientId, final Profile profile) {
         synchronized (recipients) {
             final var recipient = recipients.get(recipientId);
-            storeRecipientLocked(recipientId, Recipient.newBuilder(recipient).withProfile(profile).build());
+            if (recipient != null) {
```
Comment (AsamK) [inline lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java:264]: The recipient should never be null here, if that happens that would be a bug elsewhere.

File: lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java [hunk 3]
```diff
@@ -459,6 +464,10 @@ private void saveLocked() {
         final var base64 = Base64.getEncoder();
         var storage = new Storage(recipients.entrySet().stream().map(pair -> {
             final var recipient = pair.getValue();
+            final var number = recipient.getAddress() == null ? null : recipient.getAddress().getNumber().orElse(null);
```
Comment (AsamK) [inline lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java:467]: a recipient always has an address, so it doesn't need to be checked.

**Review by AsamK — COMMENTED** [review]
Thanks, the dbus changes look good.
But the additional null checks shouldn't be necessary.

**Comment by JtheSaw** [comment]
> But the additional null checks shouldn't be necessary.

Then there must be a bug before that. 
When the deleted account is typing a new message after deletion it seems that the daemon needs a few seconds to resolve its address and profile. Before that the checked variables above `recipient`, `recipient.getAddress()` and `profile` turn out to be null and this exception is thrown multiple times:
```
Envelope from: unknown source (device: 0)
2021-12-03T07:29:47.396528400Z Timestamp: 1638516583669 (2021-12-03T07:29:43.669Z)
2021-12-03T07:29:47.397345800Z Server timestamps: received: 1638516583748 (2021-12-03T07:29:43.748Z) delivered: 1638516583749 (2021-12-03T07:29:43.749Z)
2021-12-03T07:29:47.397787200Z Sent by unidentified/sealed sender
2021-12-03T07:29:47.398719300Z Exception: org.whispersystems.libsignal.InvalidMessageException: invalid message Message decryption failed (ProtocolInvalidMessageException)
```

Any idea how to solve this problem?

**Comment by AsamK** [comment]
The InvalidMessageException is to be expected, because the deleteRecipient method also deletes the sessions. So the decryption of the message fails. If it was an actual message (not e.g. a typing message), signal-cli will request a message resend afterwards.
I can't reproduce the issue with null recipient ... will have to analyze this further.

**Comment by JtheSaw** [comment]
If you comment out the change in `ProfileHelper.java` you should be able to reproduce the null exception.

**Comment by JtheSaw** [comment]
I think the problem is that `account.getRecipientStore().resolveRecipient(e.getSender())` in `IncomingMessageHandler.handleEnvelope` can't resolve the number and assigns a new ID to the user. So `recipients.get(recipientId)` in `RecipientStore.storeProfile` can't find any entry for this new ID and returns null. What do you think?

**Comment by AsamK** [comment]
When a new ID is assigned that ID is also persisted in the recipient store.1
I've found and fixed another bug that could cause this on master. Can you rebase and check if that's solves then null pointer exception?

File: lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java (hunk 2 above)
Comment (JtheSaw) [reply lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java:264]: I suppose this check is still needed even with the latest commit cb346fdec4ba5b5a340958699fef31aa7b2a2df5. The recipient is not added to `recipients` before `storeProfile()` is called the first time a message is received by a recipient that was deleted.

Steps to reproduce:
1. Run daemon
2. Regular messaging between a recipient account and the daemon account
3. Delete recipient with DBus method `deleteRecipient()`
4. Re-contact daemon account from deleted recipient account
5. Nullpointer exception stops daemon from working: 
```
Exception in thread "Thread-0" java.lang.NullPointerException: Cannot invoke "org.asamk.signal.manager.storage.recipients.Recipient.getRecipientId()" because "copy" is null
	at org.asamk.signal.manager.storage.recipients.Recipient.newBuilder(Recipient.java:53)
	at org.asamk.signal.manager.storage.recipients.RecipientStore.storeProfile(RecipientStore.java:264)
	at org.asamk.signal.manager.helper.ProfileHelper.getRecipientProfile(ProfileHelper.java:197)
	at org.asamk.signal.manager.helper.ProfileHelper.getRecipientProfile(ProfileHelper.java:65)
	at org.asamk.signal.manager.ManagerImpl.getRecipientProfile(ManagerImpl.java:515)
	at org.asamk.signal.manager.helper.IncomingMessageHandler.handleEnvelope(IncomingMessageHandler.java:147)
	at org.asamk.signal.manager.ManagerImpl.receiveMessagesInternal(ManagerImpl.java:1184)
	at org.asamk.signal.manager.ManagerImpl.lambda$startReceiveThreadIfRequired$11(ManagerImpl.java:1007)
	at java.base/java.lang.Thread.run(Thread.java:833)
```

Comment (JtheSaw) [reply lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java:264]: I think that the problem is still related to the ID somehow. I created some more logging for debugging below. The `IncomingMessageHandler` tries to looks up a recipient with the ID `104` but the number is already assigned to the ID `100`. (Numbers and UUIDs were hidden for privacy reasons.)

```
INFO IncomingMessageHandler - getSender:+49123456789
INFO IncomingMessageHandler - SenderID:104
INFO RecipientStore - ALL RECIPIENTS:
INFO RecipientStore - ID: RecipientId[id=100] Number: Optional[+49123456789] UUID: Optional[XXXXXX-XXX-XXXXX-XXXXXX]
INFO RecipientStore - ID: RecipientId[id=101] Number: Optional[+49...........] UUID: Optional[...............................]
INFO RecipientStore - ID: RecipientId[id=26] Number: Optional[+49...........] UUID: Optional[...............................]
```

Comment (JtheSaw) [reply lib/src/main/java/org/asamk/signal/manager/storage/recipients/RecipientStore.java:264]: That looks good, in the verbose logs I can see that the correct ID is used now.

> The InvalidMessageException is to be expected, because the deleteRecipient method also deletes the sessions. So the decryption of the message fails. If it was an actual message (not e.g. a typing message), signal-cli will request a message resend afterwards.

Still I get the expected `InvalidMessageException` and also I see that you are queuing a renew session action. The only problem is that for the deleted account it looks like the message is never received by the daemon account and also the `MessageReceived` event is not triggered. Is there a chance to fix this?

Here are my logs in case you need it:
<details>
  <summary>Verbose logs</summary>
<p>

```
2021-12-07T07:55:56.847Z [DBus Worker Thread-2] DEBUG org.asamk.signal.manager.storage.recipients.RecipientStore - Deleting recipient data for RecipientId[id=101]
2021-12-07T07:55:56.853176300Z 2021-12-07T07:55:56.852Z [DBus Worker Thread-2] DEBUG org.freedesktop.dbus.messages.MethodReturn - Creating message with serial 32
2021-12-07T07:55:56.853245400Z 2021-12-07T07:55:56.852Z [DBus Worker Thread-2] DEBUG org.freedesktop.dbus.messages.MethodReturn - Appending sig: yyyy data: [108, 2, 0, 1]
2021-12-07T07:55:56.853263200Z 2021-12-07T07:55:56.852Z [DBus Worker Thread-2] DEBUG org.freedesktop.dbus.messages.MethodReturn - Appending sig: ua(yv) data: [32, [[5, [u, 4]], [6, [s, :1.2]]]]
2021-12-07T07:55:56.854455300Z 2021-12-07T07:55:56.853Z [DBus Sender Thread-1] DEBUG org.freedesktop.dbus.spi.message.OutputStreamMessageWriter - <= MethodReturn(0,32) { Reply Serial=>4, Destination=>:1.2 } { }
2021-12-07T07:56:03.537414400Z 2021-12-07T07:56:03.535Z [Thread-0] DEBUG org.asamk.signal.manager.ManagerImpl - New message received from server
2021-12-07T07:56:03.537454100Z 2021-12-07T07:56:03.536Z [Thread-0] INFO LibSignal - [libsignal-client]: rust/protocol/src/sealed_sender.rs:441: deserialized UnidentifiedSenderMessageContent from XXXXXXX-XXXXX-XXXXX-XXXXXXX.1 with type Whisper
2021-12-07T07:56:03.542324000Z 2021-12-07T07:56:03.540Z [Thread-0] ERROR LibSignal - [libsignal-client]: rust/protocol/src/session_cipher.rs:474: No valid session for recipient: XXXXXXX-XXXXX-XXXXX-XXXXXXX.1, (no current session state), number of previous states: 0
2021-12-07T07:56:03.542369400Z 2021-12-07T07:56:03.540Z [Thread-0] ERROR LibSignal - [libsignal-client]: rust/protocol/src/session_cipher.rs:480: Message from XXXXXXX-XXXXX-XXXXX-XXXXXXX.1 failed to decrypt; sender ratchet public key XXXXXXXXXXXXXXXXXXXXXXXXX message counter 4
2021-12-07T07:56:03.542396700Z No current session
2021-12-07T07:56:03.544456100Z 2021-12-07T07:56:03.540Z [Thread-0] DEBUG org.asamk.signal.manager.storage.recipients.RecipientStore - Got new recipient, both uuid and number are unknown
2021-12-07T07:56:03.544788600Z 2021-12-07T07:56:03.540Z [Thread-0] DEBUG org.asamk.signal.manager.storage.recipients.RecipientStore - Adding new recipient RecipientId[id=124] with address RecipientAddress[uuid=Optional.empty, number=Optional[+49123456789]]
2021-12-07T07:56:03.548222300Z 2021-12-07T07:56:03.547Z [Thread-0] DEBUG org.asamk.signal.manager.helper.UnidentifiedAccessHelper - Using normal sender certificate for message to RecipientId[id=124]
2021-12-07T07:56:06.529264900Z 2021-12-07T07:56:06.528Z [Thread-0] DEBUG org.asamk.signal.manager.storage.recipients.RecipientStore - Got separate recipients for high trust number RecipientId[id=124] and uuid RecipientId[id=101], need to merge them
2021-12-07T07:56:06.662906500Z 2021-12-07T07:56:06.661Z [Thread-0] DEBUG org.asamk.signal.manager.storage.recipients.RecipientStore - Using 101 instead of 124, because recipients have been merged
2021-12-07T07:56:06.665261300Z 2021-12-07T07:56:06.664Z [Thread-0] DEBUG org.asamk.signal.manager.storage.identities.IdentityKeyStore - Storing new identity for recipient RecipientId[id=101] with trust TRUSTED_UNVERIFIED
2021-12-07T07:56:06.691109900Z 2021-12-07T07:56:06.684Z [Thread-0] DEBUG org.asamk.signal.manager.helper.IncomingMessageHandler - Received invalid message, queuing renew session action.
2021-12-07T07:56:06.691161600Z Envelope from: unknown source (device: 0)
2021-12-07T07:56:06.691188700Z Timestamp: 1638863764167 (2021-12-07T07:56:04.167Z)
2021-12-07T07:56:06.691218700Z Server timestamps: received: 1638863763442 (2021-12-07T07:56:03.442Z) delivered: 1638863763438 (2021-12-07T07:56:03.438Z)
2021-12-07T07:56:06.691244000Z Sent by unidentified/sealed sender
2021-12-07T07:56:06.691268600Z Exception: org.whispersystems.libsignal.InvalidMessageException: invalid message Message decryption failed (ProtocolInvalidMessageException)
2021-12-07T07:56:06.691308600Z 
2021-12-07T07:56:06.691351800Z 2021-12-07T07:56:06.685Z [Thread-0] DEBUG org.asamk.signal.manager.ManagerImpl - Handling message actions
2021-12-07T07:56:06.694767100Z 2021-12-07T07:56:06.693Z [Thread-0] DEBUG org.asamk.signal.manager.helper.UnidentifiedAccessHelper - Using normal sender certificate for message to RecipientId[id=101]
2021-12-07T07:56:06.808485100Z 2021-12-07T07:56:06.808Z [Thread-0] INFO LibSignal - [libsignal-client]: rust/protocol/src/session.rs:178: set_unacknowledged_pre_key_message for: XXXXXXX-XXXXX-XXXXX-XXXXXXX.1 with preKeyId: 12721785
2021-12-07T07:56:06.809131900Z 2021-12-07T07:56:06.808Z [Thread-0] INFO LibSignal - [libsignal-client]: rust/protocol/src/state/session.rs:581: Skipping archive, current session state is fresh
2021-12-07T07:56:06.814418300Z 2021-12-07T07:56:06.813Z [Thread-0] INFO LibSignal - [libsignal-client]: rust/protocol/src/session_cipher.rs:51: Building PreKeyWhisperMessage for: XXXXXXX-XXXXX-XXXXX-XXXXXXX.1 with preKeyId: 12721785
2021-12-07T07:56:06.933684500Z 2021-12-07T07:56:06.933Z [Thread-0] DEBUG org.asamk.signal.manager.ManagerImpl - Checking for new message from server
2021-12-07T07:56:07.180422700Z 2021-12-07T07:56:07.178Z [Thread-0] DEBUG org.asamk.signal.manager.ManagerImpl - New message received from server
2021-12-07T07:56:07.180464200Z Envelope from: +49123456789 (device: 1)
2021-12-07T07:56:07.180876900Z Timestamp: 1638863766693 (2021-12-07T07:56:06.693Z)
2021-12-07T07:56:07.180936000Z Server timestamps: received: 1638863767053 (2021-12-07T07:56:07.053Z) delivered: 1638863767064 (2021-12-07T07:56:07.064Z)
2021-12-07T07:56:07.181092200Z 
2021-12-07T07:56:07.181125400Z 2021-12-07T07:56:07.179Z [Thread-0] DEBUG org.asamk.signal.manager.ManagerImpl - Handling message actions
2021-12-07T07:56:07.184960100Z 2021-12-07T07:56:07.184Z [Thread-0] DEBUG org.asamk.signal.manager.ManagerImpl - Checking for new message from server
```

</p>
</details>
